third_party/caffe/Dockerfile

```diff
@@ -35,8 +35,13 @@ ENV CLONE_TAG=1.0
 # https://github.com/pypa/pip/issues/5599
 RUN git clone -b ${CLONE_TAG} --depth 1 https://github.com/BVLC/caffe.git . && \
     python -m pip install --upgrade pip && \
-    cd python && for req in $(cat requirements.txt) pydot; do pip install $req; done && cd .. && \
-    mkdir build && cd build && \
+    cd python && for req in $(cat requirements.txt) pydot; do pip install $req; done && cd ..
+
+COPY upsample.patch .
```

Comment:
Sorry for the late review.
Thanks for your wonderful code, but as you know, we can not modify the caffe code and maintain a nonstandard version, perhaps we can make this pull request stay here and help the others, but it can not be merged.

Reply:
What about keeping the original Dockerfile as it was and create a `Dockerfile-upsample` that contains the patched version?

Reply:
@gasgallo I read the code and had a question, why did you realize a `ResizeNearestNeighbor` op in caffe like in tensorflow? or you could realize a `UpSample` op like in ONNX, in that cases, there are no need to change the mace's code?

Reply:
@lu229 can you elaborate? I'm not sure I understand what you mean

Reply:
@gasgallo MACE support the `UpSample` operator for ONNX, and support `ResizeNearestNeighbor` operator for Tensorflow, What I means is that, If you simulated `UpSample` op in CAFFE like in the ONNX, or simulated `ResizeNearestNeighbor` op in CAFFE likein  the Tensorflow, there is no need to change MACE's code. I think perhaps this is a better selection?

Reply:
@lu229 I agree, it would be better if there's a way to do as you say. ONNX implementation uses `_graph_shapes_dict` object to retrieve output size, but there's no such object available in `caffe_converter` and I didn't find a similar one. Are the output shapes pre-computed before running `convert_ops()`?

Reply:
@gasgallo Sorry for the late reply, yes the output shapes can be pre-computed, you can renfer to the code in `tools/python/transform/shape_inference.py`, which is used in `caffe_converter.py' to infer the output shape.

Reply:
@lu229 but shape_inference is called after convert_ops, we need the output shapes to be computed first, as it happens in onnx_converter. And I don't think calling shape_inference before convert_ops will work

Reply:
@gasgallo Perhaps I did not understand your problem, why dose you need calling shape_inference before convert_ops? The shape inferences in ONNX and caffe and tensorflow are different, I think if you add the op in caffe, perhaps you need add it as the same the other caffe operators?

Reply:
@lu229 because current caffe upsample layer is like onnx one, that provides a scale factor, therefore we need to know upsample input size to calculate its output size that is required by MACE resizenearestneighbor op.

The other option is that caffe upsample layer directly provides output shape, like happens in tf.image.resize_nearest_neighbor.

I was wondering if the first solution would work? If not, I will consider the second option.

Reply:
@gasgallo Perhaps I still did not understand your problem, but I can try to reply. Yes you can refer the `infer_shape_conv_pool_shape` function in `tools/python/transform/shape_inference.py`, which is used to infer the output shape of `conv` operator, when you infer the `Upsample`'s output shape, you can get the input shape by the `_output_shape_cache` variable.

Reply:
@lu229 sorry for confusion, I'll write more in detail right now.

My CAFFE upsample op is defined in a similar way as ONNX upsample op, that takes a `scale` parameter, so we don't know output shape a priori. For MACE we need to know output shape when calling `self.add_tensor`.

This works in `onnx_converter.py` because we can use `self._graph_shapes_dict` object that contains all pre-computed shapes, and it is created in `self.extract_shape_info` **before** calling `self.convert_ops`.

This is `convert_upsample` in `onnx_converter.py`:
<pre><code>    def convert_upsample(self, node):
        op = self.convert_general_op(node)
        del op.input[1:]  # cut all unnecessary inputs (onnx>=1.5)

        <b>output_size = self._graph_shapes_dict[op.output[0]]</b>
        output_size = np.array(output_size[-2:]).astype(np.int32)
        if node.attrs['mode'] == 'nearest':
            op.type = MaceOp.ResizeNearestNeighbor.name
            size_tensor_name = op.name + ":size"
            self.add_tensor(size_tensor_name, output_size.shape,
                            mace_pb2.DT_INT32, output_size)
            op.input.append(size_tensor_name)
        else:
            op.type = MaceOp.ResizeBilinear.name
            size_arg = op.arg.add()
            size_arg.name = MaceKeyword.mace_resize_size_str
            size_arg.ints.extend(output_size.tolist())

        align_corners_arg = op.arg.add()
        align_corners_arg.name = MaceKeyword.mace_align_corners_str
        align_corners_arg.i = node.attrs.get('align_corners', 0)
</code></pre>

And this is `run()` in `onnx_converter.py`:
<pre><code>    def run(self):
        graph_def = self._onnx_model.graph
        <b>self.extract_shape_info(graph_def)</b>
        self.convert_tensors(graph_def)
        <b>self.convert_ops(graph_def)</b>
        return self._mace_net_def
</code></pre>

In `caffe_converter.py` instead, `shape_inferer.run()` is called **after** `self.convert_ops`, therefore all output shapes are not available during ops conversion, so I cannot compute output shape of upsample op.

This is `run()` in `caffe_converter.py`:
<pre><code>    def run(self):
        <b>self.convert_ops()</b>
        shape_inferer = shape_inference.ShapeInference(
            self._mace_net_def,
            self._option.input_nodes.values())
        <b>shape_inferer.run()</b>
        self.replace_output_tensor_name()
        return self._mace_net_def
</code></pre>

And `shape_inferer.run()` cannot be called before `self.convert_ops` because it requires `self._mace_net_def` to be already completely defined (that happens in `self.convert_ops`).

@lu229 Let me know if this makes my point clear.

Reply:
@gasgallo Thanks for your detailed description, I get your question, yes you need to define the upsample as same as the other `Caffe` operators, for the shape inference in `Caffe` and `ONNX` is different.